Objects/memoryobject.c

```diff
@@ -3160,6 +3160,92 @@ static PyMethodDef memory_methods[] = {
     {NULL,          NULL}
 };
 
+/**************************************************************************/
+/*                          Memoryview Iterator                           */
+/**************************************************************************/
+
+typedef struct {
+    PyObject_HEAD
+    Py_ssize_t it_index;
+    PyMemoryViewObject *it_seq; // Set to NULL when iterator is exhausted
+} memoryiterobject;
+
+
+static void memoryiter_dealloc(memoryiterobject *);
+static int memoryiter_traverse(memoryiterobject * , visitproc, void *);
+static PyObject * memoryiter_next(memoryiterobject *);
+static PyObject *memory_iter(PyObject *);
+
+
+static PyTypeObject PyMemoryIter_Type = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "memory_iterator",
+    .tp_basicsize = sizeof(memoryiterobject),
+    // methods
+    .tp_dealloc = (destructor)memoryiter_dealloc,
+    .tp_getattro = PyObject_GenericGetAttr,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_traverse = (traverseproc)memoryiter_traverse,
+    .tp_iter = PyObject_SelfIter,
+    .tp_iternext = (iternextfunc)memoryiter_next,
+};
+
+
+static void
+memoryiter_dealloc(memoryiterobject *it)
+{
+  _PyObject_GC_UNTRACK(it);
+  Py_XDECREF(it->it_seq);
+  PyObject_GC_Del(it);
+}
+
+static int
+memoryiter_traverse(memoryiterobject *it, visitproc visit, void *arg)
+{
+  Py_VISIT(it->it_seq);
+  return 0;
+}
+
+static PyObject *
+memoryiter_next(memoryiterobject *it)
+{
+    PyMemoryViewObject *seq;
+
+    assert(it != NULL);
+    seq = it->it_seq;
+    if (seq == NULL)
+      return NULL;
+    assert (PyMemoryView_Check(seq));
+
+    if (it->it_index < PyMemoryView_GET_SIZE(seq)){
+        return PyObject_GetItem(
+                (PyObject * )seq, PyLong_FromSize_t(it->it_index++));
+    }
+
+    it->it_seq = NULL;
```

Comment:
Try inserting `PyErr_Clear();` before this line. (See iterobject.c.)

If that fixes things, I think the IndexError you got was a leftover exception from a previous call (an unfortunate problem with Python exception handling in C).

If that doesn't fix things, let me know and I'll debug it for you.

Reply:
I tried to call PyErr_Clear() but it doesn't change the result.

An interesting thing I noticed, which perhaps will help understanding why the wrong exception is being raised, is the following:

```python
a = b"abc" 
b = memoryview(a) 
c = iter(b)
next(it)
next(it)
next(it)
next(it) # Correctly raises StopIteration
```
```python
a = b"ab" 
b = memoryview(a)
c = iter(b)
next(it)
next(it)
next(it) # raises IndexError
```
Basically it raises IndexError when the bytes (or bytearray) object has length < 3, while if it the object passed to memoryview has length >= 3 it properly raises StopIteration.

I will continue working on it.

Reply:
Here are some strategies to get a hint from the working of the examples.

Does it go differently when you exit and restart python in between trying the two examples? What happens if you reverse them? (2 first, 3 last). What happens for other sizes? Is there a pattern?

Reply:
Thank you Guido. I will report back after doing the analysis you recommend.

Reply:
After a first analysis, I noticed that the iteration seems to be limited to 3 values, regardless of the size:

```python
  x = b"abcde"    
  m = memoryview(x)    
  it = iter(m)    
  next(it)    
  next(it)    
  next(it)    
  next(it) # Incorrectly raises StopIteration
```
I haven't understood yet why it is so but I will continue working on it.

Reply:
If you print the value of `PyMemoryView_GET_SIZE(seq)` you'll find that it's always 3, which is the cause of the problem.

How do you print it? I added this to the code, just before the line where it's used:
```
    printf("size=%ld\n", PyMemoryView_GET_SIZE(seq));
```
(Hopefully your books have covered `printf()`. The `%ld` format code is needed because the GET_SIZE macro returns a ssize_t, which is an integer type that's larger than the default int. The output with `%d` is the same; you will probably just get a compiler warning, but for some systems it may break.)

Now what's wrong with the GET_SIZE calculation? You added a macro, PyMemoryView_GET_SIZE to calculate the size. But if you look at memory_length you'll see that the memoryview object itself calculates its size quite differently. Apparently the GET_SIZE macro accesses some random field of the struct that isn't set to the length of the first dimension.

How do you calculate the correct size? Use PyObject_Size (or call memory_length directly).

Hopefully this gets you unstuck.

PS. I was wondering about the value 3. Why 3? The answer is in memory_alloc, where the expression `3*ndim` occurs. Why is it doing this? I have no idea. Probably not a good use of our time to try to understand this.

PPSS. You know how to use tags in vim, right? The Python Makefile has a handy target, `make tags` which creates a tags file for you. If you've never heard of this before, try Googling. It's super helpful for jumping around in the code, especially across files.

Reply:
It totally makes sense now. I shot myself on the leg with that macro and can confirm that it works as supposed if I call memory_length(seq) instead.
On that note I tried to return memory_item(seq, it->it_index++) instead of calling PyObject_GetItem and it seems to be working properly. Do you think we should use the former or the latter function? The former seems to me that it could eventually be faster by cutting a level of indirection (proper tests would be necessary to confirm) although currently seems to lack support for multi-dimensional arrays. 
What are your thoughts on it?

Reply:
Yeah, that’s a good idea. Later we can optimize further. Now to get the tests passing...

Reply:
Ok. I will push a new commit with the changes we spoke and will look into the tests.